examples/local_damage.py

```diff
@@ -0,0 +1,249 @@
+    """
+    Creates something like
+     ly +---------+
+        |    ^y   |
+        |    |    |
+        |    0->--+
+        |    | x
+        |    |
+    -ly +----+
+       -lx       lx
+        out of triangles
+    """
+    mesh = df.Mesh()
+
+    e = df.MeshEditor()
+    e.open(mesh, "triangle", 2, 2)
+    e.init_vertices(8)
+    e.add_vertex(0, [0, 0])
+    e.add_vertex(1, [lx, 0])
+    e.add_vertex(2, [lx, ly])
+    e.add_vertex(3, [0, ly])
+    e.add_vertex(4, [-lx, ly])
+    e.add_vertex(5, [-lx, 0])
+    e.add_vertex(6, [-lx, -ly])
+    e.add_vertex(7, [0, -ly])
+
+    e.init_cells(6)
+    e.add_cell(0, [0, 1, 3])
+    e.add_cell(1, [1, 2, 3])
+    e.add_cell(2, [0, 3, 5])
+    e.add_cell(3, [5, 3, 4])
+    e.add_cell(4, [0, 5, 7])
+    e.add_cell(5, [7, 5, 6])
+
+    e.close
+    mesh.order()
+
+    for _ in range(refinement):
+        mesh = df.refine(mesh)
+
+    if show:
+        df.plot(mesh)
+        import matplotlib.pyplot as plt
+        plt.show()
+
+
+    return mesh
+
+
+def max(a, b):
+    return 0.5 * (a + b + abs(a - b))
+
+class Mat:
+    E= 20000
+    nu = 0.20
+    ft = 4.
+    Gf = 0.2
+    alpha = 0.99
+
+def omega(k):
+    k0 = Mat.ft / Mat.E
+    beta = Mat.ft / Mat.Gf
+    exponent = beta * (k0 - k)
+    omega = 1.0 - k0 / k* (1.0 - Mat.alpha + Mat.alpha * df.exp(exponent))
+    return df.conditional(df.lt(k, k0), 0, omega)
+
+class ModifiedMises:
+    """ Modified von Mises equivalent strain, see
+    de Vree et al., 1995, "Comparison of Nonlocal Approaches in
+    Continuum Damage Mechanics"
+
+    Invariants from https://en.wikipedia.org/wiki/Cauchy_stress_tensor
+
+    The members T1, T2, T3 correspond to the term 1,2,3 in the equation
+    """
+    def __init__(self, k, nu):
+        self.k = k
+        self.nu = nu
+        self.T1 = (k - 1) / (2. * k * (1. - 2. * nu))
+        self.T2 = (k - 1) / (1. - 2. * nu)
+        self.T3 = 12. * k / ((1. + nu) * (1. + nu))
+
+    def __call__(self, eps):
+        I1 = df.tr(eps)
+        J2 = 0.5 * df.tr(df.dot(eps, eps)) - 1. / 6. * df.tr(eps) * df.tr(eps)
+
+        A = (self.T2 * I1) ** 2 + self.T3 * J2
+        A_pos = max(A, 1.e-14)
+
+        return self.T1 * I1 + df.sqrt(A_pos) / (2. * self.k)
+
+class Problem:
+    def __init__(self, mat, mesh, order=1):
+        self.mat = mat
+        self.mesh = mesh
+
+        self.W = df.VectorFunctionSpace(mesh, "P", order)
+        self.W_k = df.FunctionSpace(mesh, "P", order)
+
+        self.d = df.Function(self.W)
+        # self.v = df.TestFunction(self.W)
+
+        self.k = df.Function(self.W_k)
+
+        self.eeq = ModifiedMises(10, mat.nu)
+
+    def eps(self, d=None):
+        if d is None:
+            d = self.d
+        return df.sym(df.grad(d))
+    
+    def sigma(self):
+        eps, E, nu = self.eps(), self.mat.E, self.mat.nu
+        lmbda = E * nu / (1. + nu) / (1. - 2. * nu)
+        mu = E / 2. / (1. + nu)
+
+        return 2 * mu * eps + lmbda * df.tr(eps) * df.Identity(2)
```

Comment:
This seems like the only place where the dimension of the `Problem` class is hardcoded. I know this is only an example, but maybe you can do it dimension-independent?